connect-inject/endpoints_controller.go

```diff
@@ -392,24 +386,23 @@ func getConsulHealthCheckID(pod corev1.Pod, serviceID string) string {
 	return fmt.Sprintf("%s/%s/kubernetes-health-check", pod.Namespace, serviceID)
 }
 
-// getReadyStatusAndReason returns the formatted status string to pass to Consul based on the
-// ready state of the pod along with the reason message which will be passed into the Notes
-// field of the Consul health check.
-func getReadyStatusAndReason(pod corev1.Pod) (string, string, error) {
-	for _, cond := range pod.Status.Conditions {
-		var consulStatus, reason string
-		if cond.Type == corev1.PodReady {
-			if cond.Status != corev1.ConditionTrue {
-				consulStatus = api.HealthCritical
-				reason = cond.Message
-			} else {
-				consulStatus = api.HealthPassing
-				reason = kubernetesSuccessReasonMsg
-			}
-			return consulStatus, reason, nil
-		}
+// getHealthCheckStatusReason takes an endpoint address and Consul's health check status (either passing or critical)
+// and returns the reason message. It assumes that the address has a TargetRef pointing to a pod,
+// otherwise it returns an empty string.
+func getHealthCheckStatusReason(address corev1.EndpointAddress, healthCheckStatus string) string {
+	if healthCheckStatus == api.HealthPassing {
+		return kubernetesSuccessReasonMsg
+	}
+
+	// Get the pod name from the address if it points to a pod,
+	// otherwise use the IP of the endpoint address in the reason message.
+	if address.TargetRef != nil && address.TargetRef.Kind == "Pod" {
+		return fmt.Sprintf("Pod \"%s/%s\" is not ready", address.TargetRef.Namespace, address.TargetRef.Name)
+	} else {
+		// If TargetRef is not a pod, the endpoints controller will ignore it and not call this function,
+		// and so we don't need to worry about the reason message in that case.
+		return ""
```

Comment:
If this is a case we shouldn't hit should this say something like "error: targetRef should not be null"?

Reply:
Yeah, I was on the fence on whether to return an error here, but could certainly do so! It will look cleaner from an API perspective even though it's a private function.

Reply:
Rather than an error I was thinking you could literally return that string. But error would work too.

Reply:
An error or the string literal suggested above sounds good to me

Reply:
I've refactored it to instead take pod name and namespace as args to the function. I think it makes more sense since we're using the address only to get the name and the namespace anyways.